From a603769010abf8c1bede91af46e4945314e4aa4a Mon Sep 17 00:00:00 2001
From: Kenny2github <kenny2minecraft@gmail.com>
Date: Fri, 20 Nov 2020 03:07:16 +0800
Subject: [PATCH] Change behavior of PUT /verify/{}

* Instead of returning the same code, which someone else could hijack,
  generate a new code on repeated PUT to preserve integrity.
* Update docs to reflect this new behavior
---
 backend/db.py       | 11 ++++++-----
 docs/docs.json      |  2 +-
 docs/reference.html |  2 +-
 3 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/backend/db.py b/backend/db.py
index 0b4086a..4cf7031 100644
--- a/backend/db.py
+++ b/backend/db.py
@@ -162,11 +162,6 @@ async def start_verification(self, client_id, username):
             await self.db.execute('SELECT code FROM scratchverifier_usage WHERE \
 client_id=? AND username=?', (client_id, username))
             row = await self.db.fetchone()
-        if row is not None:
-            await self.db.execute('UPDATE scratchverifier_usage SET expiry=? \
-WHERE client_id=? AND username=? AND code=?', (int(time.time()) + VERIFY_EXPIRY,
-                                               client_id, username, row[0]))
-            return row[0]
         code = sha256(
             str(client_id).encode()
             + str(time.time()).encode()
@@ -174,6 +169,12 @@ async def start_verification(self, client_id, username):
             + token_bytes()
         # 0->A, 1->B, etc, to avoid Scratch's phone number censor
         ).hexdigest().translate({ord('0') + i: ord('A') + i for i in range(10)})
+        if row is not None:
+            await self.db.execute(
+                'UPDATE scratchverifier_usage SET expiry=?, code=? \
+WHERE client_id=? AND username=?', (int(time.time()) + VERIFY_EXPIRY,
+                                    code, client_id, username))
+            return code
         await self.db.execute('INSERT INTO scratchverifier_usage (client_id, \
 code, username, expiry) VALUES (?, ?, ?, ?)', (client_id, code, username,
                                int(time.time() + VERIFY_EXPIRY)))
diff --git a/docs/docs.json b/docs/docs.json
index 55edd31..446b37f 100644
--- a/docs/docs.json
+++ b/docs/docs.json
@@ -59,7 +59,7 @@
     {
       "type": "endpoint",
       "name": "Start/Renew Verification",
-      "desc": "Request a new verification code for a user. Only one code per user per client - if this endpoint is used again before the <a href=\"#finish-verification-endpoint\">Finish Verification</a> endpoint is used, this will instead renew the 30-minute expiry on the code and return the original code.",
+      "desc": "Request a new verification code for a user. Only one code per user per client - if this endpoint is used again before the <a href=\"#finish-verification-endpoint\">Finish Verification</a> endpoint is used, this will generate a new code and reset the 30-minute expiry, returning the new code instead.",
       "method": "PUT",
       "path": "/verify/{username}",
       "params": {
diff --git a/docs/reference.html b/docs/reference.html
index 9808f5d..8c2ce16 100644
--- a/docs/reference.html
+++ b/docs/reference.html
@@ -21,7 +21,7 @@ <h2 id="authorization">Authorization</h2>
 <h2 id="api-endpoints">API Endpoints</h2>
 <p>The simplicity of this API is such that there are only three total endpoints for its ultimate purpose.</p>
 <h3 id="start/renew-verification-endpoint">Start/Renew Verification Endpoint</h3>
-<p>Request a new verification code for a user. Only one code per user per client - if this endpoint is used again before the <a href="#finish-verification-endpoint">Finish Verification</a> endpoint is used, this will instead renew the 30-minute expiry on the code and return the original code.</p>
+<p>Request a new verification code for a user. Only one code per user per client - if this endpoint is used again before the <a href="#finish-verification-endpoint">Finish Verification</a> endpoint is used, this will generate a new code and reset the 30-minute expiry, returning the new code instead.</p>
 <div class="endpoint">
 <div class="method-path auth-needed" onclick="showOrHide(this)"><span class="method">PUT</span> <code>/verify/<span class="param">{username}</span></code> <a href="#authorization"><img src="https://image.flaticon.com/icons/png/512/61/61457.png" title="Authorization necessary" /></a></div>
 <div style="display: none">
